packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1625 @@
+                        KGH.safeTransferFrom(address(this), msg.sender, tokenId[tokenIdIndex]);
+
+                        unchecked {
+                            ++tokenIdIndex;
+                        }
+                    }
+
+                    delete kghDelegator.pendingShares[requestTimes[i]];
+                    delete kghDelegator.pendingKghIds[requestTimes[i]];
+                    delete kghDelegator.undelegateRequestTimes[i];
+                } else {
+                    break;
+                }
+            }
+
+            unchecked {
+                --index;
+            }
+        }
+
+        unchecked {
+            asset.totalPendingAssets -= kroAssetsToUndelegate;
+            asset.totalPendingKroShares -= kroSharesToUndelegate;
+            asset.totalPendingBoostedRewards -= kghAssetsToUndelegate;
+            asset.totalPendingKghShares -= kghSharesToUndelegate;
+        }
+
+        uint128 assetsToUndelegate = kroAssetsToUndelegate + kghAssetsToUndelegate;
+
+        SafeERC20.safeTransfer(ASSET_TOKEN, msg.sender, assetsToUndelegate);
+
+        emit KghUndelegationFinalized(validator, msg.sender, assetsToUndelegate);
+        return assetsToUndelegate;
+    }
+
+    /**
+     * @notice Finalize the reward claim of the validator.
+     */
+    function finalizeClaimValidatorReward() external {
+        Reward storage reward = _vaults[msg.sender].reward;
+        uint256[] memory requestTimes = reward.claimRequestTimes;
+
+        uint128 assetsToWithdraw;
+        for (uint256 index = requestTimes.length; index > 0; ) {
+            uint256 i = index - 1;
+            if (requestTimes[i] + UNDELEGATION_PERIOD <= block.timestamp) {
+                if (reward.pendingValidatorRewards[requestTimes[i]] > 0) {
+                    assetsToWithdraw += reward.pendingValidatorRewards[requestTimes[i]];
+
+                    delete reward.pendingValidatorRewards[requestTimes[i]];
+                    delete reward.claimRequestTimes[i];
+                } else {
+                    break;
+                }
+            }
+
+            unchecked {
+                --index;
+            }
+        }
+
+        // To prevent the underflow of the totalPendingValidatorRewards when the validator is slashed.
+        if (reward.totalPendingValidatorRewards < assetsToWithdraw) {
+            assetsToWithdraw = reward.totalPendingValidatorRewards;
+        }
+
+        reward.totalPendingValidatorRewards -= assetsToWithdraw;
+        SafeERC20.safeTransfer(ASSET_TOKEN, msg.sender, assetsToWithdraw);
+
+        emit RewardClaimFinalized(msg.sender, assetsToWithdraw);
+    }
+
+    /**
+     * @notice Internal function to add base and boosted reward to the vaults of finalized output submitters.
+     *
+     * @return Whether the reward distribution is done at least once or not.
+     */
+    function _distributeReward() internal returns (bool) {
```

Comment:
When the reward is distributed, the real token should be transferred to this contract by L2OO contract, right?
But for now, there is no actual vault to send the reward to this contract. If so, please add TODO mark.